src/Autofac/Core/Activators/Reflection/ReflectionActivator.cs

```diff
@@ -310,33 +352,112 @@ private string GetBindingFailureMessage(BoundConstructor[] constructorBindings)
         }
     }
 
-    private void InjectProperties(object instance, IComponentContext context)
+    private void InjectProperties(object instance, IComponentContext context, BoundConstructor constructor, IEnumerable<Parameter> allParameters)
     {
-        if (_configuredProperties.Length == 0)
+        // We only need to do any injection if we have a set of property information.
+        if (_defaultFoundPropertySet is null)
         {
             return;
         }
 
-        var actualProperties = instance
-            .GetType()
-            .GetRuntimeProperties()
-            .Where(pi => pi.CanWrite)
-            .ToList();
+        // If this constructor sets all required members,
+        // and we have no configured properties, we can just jump out.
+        if (_configuredProperties.Length == 0 && constructor.SetsRequiredMembers)
+        {
+            return;
+        }
+
+        var workingSetOfProperties = (InjectablePropertyState[])_defaultFoundPropertySet.Clone();
```

Comment:
I think [`Array.Clone()` is a shallow copy](https://learn.microsoft.com/en-us/dotnet/api/system.array.clone?view=net-7.0) so the clone is going to point to the same instances of `InjectablePropertyState` as the original. Wouldn't that mean once `InjectablePropertyState.Set` is set once for a given type that it wouldn't be tried again? (I'm not setting up tests to verify this at the moment, just thinking here.)

Like...

```c#
var builder = new ContainerBuilder();
builder.RegisterType<ThingWithRequiredProperties>();
buidler.RegisterType<ValueForRequiredProperty>();
var container = builder.Build();

// This one would set the required properties.
_ = container.Resolve<ThingWithRequiredProperties>();

// This one would think they're already set.
_ = container.Resolve<ThingWithRequiredProperties>();
```

Or maybe that's the magic of the `struct` working there, that the shallow copy ends up _being_ a deep copy because it's not a reference type?

Reply:
Correct, it's the magic of `InjectablePropertyState` being a struct, it performs a deep copy (or at least copies the current values of the existing collection items). References inside that struct don't get copied.